Sources/AuthFoundation/Network/APIClient.swift

```diff
@@ -116,14 +157,60 @@ extension APIClient {
                 completion(.failure(apiError))
                 return
             }
-
+            
             do {
-                let response: APIResponse<T> = try self.validate(data: data,
-                                                                 response: response,
-                                                                 parsing: context)
-                self.didSend(request: request, received: response)
+                guard let httpResponse = response as? HTTPURLResponse else {
+                    throw APIClientError.invalidResponse
+                }
                 
-                completion(.success(response))
+                let rateInfo = ApiRateLimit(with: httpResponse.allHeaderFields)
+                switch httpResponse.statusCode {
+                case 200..<300:
+                    let response: APIResponse<T> = try self.validate(data: data,
+                                                                     response: httpResponse,
+                                                                     rateInfo: rateInfo,
+                                                                     parsing: context)
+                    self.didSend(request: request, received: response)
+                    completion(.success(response))
+                case 429:
+                    guard let rateInfo = rateInfo else {
+                        fallthrough
+                    }
+                    let retry = state?.type ?? self.shouldRetry(request: request, rateLimit: rateInfo)
+                    switch retry {
+                    case .doNotRetry: break
+                    case .retry(let maximumRetryCount):
+                        let retryState: APIRetry.State
+                        if let state = state {
+                            retryState = state.nextState()
+                        } else {
+                            guard let requestIdHeader = requestIdHeader,
+                                  let requestId = httpResponse.allHeaderFields[requestIdHeader] as? String
+                            else {
+                                throw APIClientError.noRequestId
```

Comment:
I don't think we want to throw an exception in this case. We're mostly doing this for traceability, it's not a strict requirement for it to work (to my knowledge)

Reply:
The requestId is added to the retry requests` X-Okta-Retry-For` which is necessary from my understanding. without throwing an exception in cases where it's missing from my original request, we would not be adding that to the retry request header

Reply:
Right. I'm suggesting the `X-Okta-Retry-For` is optional, and not including it should not be considered an error.

Reply:
Okay. I think that's a good suggestion. Seems like that would be consistent with the donet implementation as well. I'll make the changes.

Reply:
I've updated this. Thanks